mysys/my_largepage.c

```diff
@@ -15,53 +15,43 @@
 
 #include "mysys_priv.h"
 
+#ifdef HAVE_SYS_MMAN_H
+#include <sys/mman.h>
+#endif
 #ifdef HAVE_LINUX_LARGE_PAGES
-
-#ifdef HAVE_SYS_IPC_H
-#include <sys/ipc.h>
+#include <linux/mman.h>
+#include <dirent.h>
+#include "my_bit.h"
 #endif
 
-#ifdef HAVE_SYS_SHM_H
-#include <sys/shm.h>
-#endif
 
-static uint my_get_large_page_size_int(void);
-static uchar* my_large_malloc_int(size_t size, myf my_flags);
-static my_bool my_large_free_int(uchar* ptr);
+static inline my_bool my_is_2pow(size_t n) { return !((n) & ((n) - 1)); }
 
-/* Gets the size of large pages from the OS */
+static uchar* my_large_malloc_int(size_t *size, myf my_flags);
+static my_bool my_large_free_int(void *ptr, size_t size);
 
-uint my_get_large_page_size(void)
-{
-  uint size;
-  DBUG_ENTER("my_get_large_page_size");
-  
-  if (!(size = my_get_large_page_size_int()))
-    fprintf(stderr, "Warning: Failed to determine large page size\n");
-
-  DBUG_RETURN(size);
-}
+#ifdef HAVE_LARGE_PAGE_OPTION
 
 /*
   General large pages allocator.
   Tries to allocate memory from large pages pool and falls back to
   my_malloc_lock() in case of failure
 */
 
-uchar* my_large_malloc(size_t size, myf my_flags)
+uchar* my_large_malloc(size_t *size, myf my_flags)
 {
   uchar* ptr;
   DBUG_ENTER("my_large_malloc");
   
-  if (my_use_large_pages && my_large_page_size)
+  if ((ptr = my_large_malloc_int(size, my_flags)) != NULL)
   {
-    if ((ptr = my_large_malloc_int(size, my_flags)) != NULL)
-        DBUG_RETURN(ptr);
-    if (my_flags & MY_WME)
-      fprintf(stderr, "Warning: Using conventional memory pool\n");
+    MEM_UNDEFINED(ptr, *size);
```

Comment:
By fact, on every platform `my_large_malloc_int()` returns zero-filled buffer. I suggest to make it a part of function interface (mention in comment) and do not poison memory here at all.

Reply:
Sorry, it's actually unposoning happens here, OK, it doesn't matter because there is no sense in unpoisoning a zero-initialized buffer.